README.md
# shogi-rs

[![Github Actions](https://github.com/nozaq/shogi-rs/workflows/build/badge.svg)](https://github.com/nozaq/shogi-rs/actions?workflow=build)
[![Coverage Status](https://coveralls.io/repos/github/nozaq/shogi-rs/badge.svg)](https://coveralls.io/github/nozaq/shogi-rs)
[![crates.io](https://img.shields.io/crates/v/shogi.svg)](https://crates.io/crates/shogi)
[![docs.rs](https://docs.rs/shogi/badge.svg)](https://docs.rs/shogi)

A Bitboard-based shogi library in Rust. Board representation, move generation/validation and time control utilities.

[Documentation](https://docs.rs/shogi)

## Usage

A library for implementing Shogi application.

`shogi` provides a various types and implementations for representing concepts and rules in Shogi.
Most types can be created programatically while they can also be deserialized from / serialized to SFEN format.
See [USIプロトコルとは (What is the USI protocol?)](http://shogidokoro.starfree.jp/usi.html) for more detail about UCI protocol specification and SFEN format.

## Examples

```rust
use shogi::{Move, Position};
use shogi::bitboard::Factory as BBFactory;
use shogi::square::consts::*;

BBFactory::init();
let mut pos = Position::new();

// Position can be set from the SFEN formatted string.
pos.set_sfen("lnsgkgsnl/1r5b1/ppppppppp/9/9/9/PPPPPPPPP/1B5R1/LNSGKGSNL b - 1").unwrap();

// You can programatically create a Move instance.
let m = Move::Normal{from: SQ_7G, to: SQ_7F, promote: false};
pos.make_move(m).unwrap();

// Move can be created from the SFEN formatted string as well.
let m = Move::from_sfen("7c7d").unwrap();
pos.make_move(m).unwrap();

// Position can be converted back to the SFEN formatted string.
assert_eq!("lnsgkgsnl/1r5b1/ppppppppp/9/9/9/PPPPPPPPP/1B5R1/LNSGKGSNL b - 1 moves 7g7f 7c7d", pos.to_sfen());
```

## Related crates

- [csa-rs](https://github.com/nozaq/csa-rs): A Shogi game serialization/deserialization library in CSA format. 
- [usi-rs](https://github.com/nozaq/usi-rs): A library to handle type-safe communication with USI-compatible shogi engines. 

## License

`shogi-rs` is licensed under the MIT license. Please read the [LICENSE](LICENSE) file in this repository for more information.


CHANGELOG.md
<a name="unreleased"></a>
## [Unreleased]


<a name="0.12.2"></a>
## [0.12.2] - 2021-12-27
### Chore
- upgrade to 2021 edition ([#45](https://github.com/nozaq/shogi-rs/issues/45))
- upgrade checkout action to v2 ([#44](https://github.com/nozaq/shogi-rs/issues/44))


<a name="0.12.1"></a>
## [0.12.1] - 2021-12-09
### Fix
- Uchifuzume check ([#41](https://github.com/nozaq/shogi-rs/issues/41))
- typo and reference link ([#40](https://github.com/nozaq/shogi-rs/issues/40))


<a name="0.12.0"></a>
## [0.12.0] - 2021-11-14
### Feat
- make Position#find_king() public ([#39](https://github.com/nozaq/shogi-rs/issues/39))


<a name="0.11.0"></a>
## [0.11.0] - 2021-10-04
### Feat
- make Position#pinned_bb public ([#36](https://github.com/nozaq/shogi-rs/issues/36))
- modernize error types ([#35](https://github.com/nozaq/shogi-rs/issues/35))


<a name="0.10.3"></a>
## [0.10.3] - 2021-10-03
### Chore
- rename `master` branch to `main` ([#33](https://github.com/nozaq/shogi-rs/issues/33))

### Fix
- add inc_time before the turn begins ([#34](https://github.com/nozaq/shogi-rs/issues/34))


<a name="0.10.2"></a>
## [0.10.2] - 2021-08-15

<a name="0.10.1"></a>
## [0.10.1] - 2021-08-15
### Fix
- broken compilation in perft ([#31](https://github.com/nozaq/shogi-rs/issues/31))


<a name="0.10.0"></a>
## [0.10.0] - 2020-11-14
### Chore
- use tarpaulin to measure coverage

### Fix
- use range expressions
- clippy warnings
- rename an unused variable
- unmaking of capturing move ([#27](https://github.com/nozaq/shogi-rs/issues/27))


<a name="0.9.0"></a>
## [0.9.0] - 2019-12-30
### Chore
- delete unnecessary panic settings
- use Github Actions

### Refactor
- fix clippy warnings and errors
- use dyn for specifying trait objects
- use ? instead of r#try


<a name="0.8.0"></a>
## [0.8.0] - 2019-01-24
### Chore
- use git-chglog to generate CHANGELOG
- remove redundant builds
- add appveyor config

### Feat
- add Position#player_bb()

### Fix
- add Cargo.toml
- Coveralls integration

### Refactor
- migrate to 2018 edition
- upgrade itertools to v0.8.0
- upgrade bitintr to v0.2


<a name="0.7.0"></a>
## [0.7.0] - 2017-05-14
### Refactor
- extract 'usi' module as an external crate


<a name="0.6.0"></a>
## [0.6.0] - 2017-03-05

<a name="0.5.0"></a>
## [0.5.0] - 2017-02-25
### Feat
- ensure Square not to return invalid values ([#7](https://github.com/nozaq/shogi-rs/issues/7))

### Fix
- check pinning status in Uchifuzume detection ([#8](https://github.com/nozaq/shogi-rs/issues/8))


<a name="0.4.0"></a>
## [0.4.0] - 2017-02-22
### Feat
- add Nyugyoku check ([#6](https://github.com/nozaq/shogi-rs/issues/6))


<a name="0.3.0"></a>
## [0.3.0] - 2017-02-19
### Feat
- add a helper type to manage time controls ([#5](https://github.com/nozaq/shogi-rs/issues/5))

### Refactor
- remove an unused variable


<a name="0.2.0"></a>
## [0.2.0] - 2017-02-19
### Chore
- temporary workaround for coverall integration ([#3](https://github.com/nozaq/shogi-rs/issues/3))

### Feat
- USI protocol helpers ([#4](https://github.com/nozaq/shogi-rs/issues/4))


<a name="0.1.0"></a>
## 0.1.0 - 2017-02-18
### Chore
- add travis.yml


[Unreleased]: https://github.com/nozaq/shogi-rs/compare/0.12.2...HEAD
[0.12.2]: https://github.com/nozaq/shogi-rs/compare/0.12.1...0.12.2
[0.12.1]: https://github.com/nozaq/shogi-rs/compare/0.12.0...0.12.1
[0.12.0]: https://github.com/nozaq/shogi-rs/compare/0.11.0...0.12.0
[0.11.0]: https://github.com/nozaq/shogi-rs/compare/0.10.3...0.11.0
[0.10.3]: https://github.com/nozaq/shogi-rs/compare/0.10.2...0.10.3
[0.10.2]: https://github.com/nozaq/shogi-rs/compare/0.10.1...0.10.2
[0.10.1]: https://github.com/nozaq/shogi-rs/compare/0.10.0...0.10.1
[0.10.0]: https://github.com/nozaq/shogi-rs/compare/0.9.0...0.10.0
[0.9.0]: https://github.com/nozaq/shogi-rs/compare/0.8.0...0.9.0
[0.8.0]: https://github.com/nozaq/shogi-rs/compare/0.7.0...0.8.0
[0.7.0]: https://github.com/nozaq/shogi-rs/compare/0.6.0...0.7.0
[0.6.0]: https://github.com/nozaq/shogi-rs/compare/0.5.0...0.6.0
[0.5.0]: https://github.com/nozaq/shogi-rs/compare/0.4.0...0.5.0
[0.4.0]: https://github.com/nozaq/shogi-rs/compare/0.3.0...0.4.0
[0.3.0]: https://github.com/nozaq/shogi-rs/compare/0.2.0...0.3.0
[0.2.0]: https://github.com/nozaq/shogi-rs/compare/0.1.0...0.2.0


src/bitboard.rs
use std::fmt;
use std::ops;
use bitintr::x86::bmi2::pext;

use Square;

#[derive(Debug, Default, Clone, Copy)]
pub struct Bitboard {
    p: [u64; 2],
}

/////////////////////////////////////////////////////////////////////////////
// Type implementation
/////////////////////////////////////////////////////////////////////////////

impl Bitboard {
    pub fn new(p1: u64, p2: u64) -> Bitboard {
        Bitboard { p: [p1, p2] }
    }

    pub fn init() {}
}

/////////////////////////////////////////////////////////////////////////////
// Operator implementations
/////////////////////////////////////////////////////////////////////////////

impl<'a> ops::Not for &'a Bitboard {
    type Output = Bitboard;

    fn not(self) -> Bitboard {
        Bitboard { p: [!self.p[0], !self.p[1]] }
    }
}

impl<'a, 'b> ops::BitAnd<&'a Bitboard> for &'b Bitboard {
    type Output = Bitboard;

    fn bitand(self, rhs: &'a Bitboard) -> Bitboard {
        Bitboard { p: [self.p[0] & rhs.p[0], self.p[1] & rhs.p[1]] }
    }
}

impl<'a> ops::BitAndAssign<&'a Bitboard> for Bitboard {
    fn bitand_assign(&mut self, rhs: &'a Bitboard) {
        self.p[0] &= rhs.p[0];
        self.p[1] &= rhs.p[1];
    }
}

impl<'a, 'b> ops::BitOr<&'a Bitboard> for &'b Bitboard {
    type Output = Bitboard;

    fn bitor(self, rhs: &'a Bitboard) -> Bitboard {
        Bitboard { p: [self.p[0] | rhs.p[0], self.p[1] | rhs.p[1]] }
    }
}

impl<'a> ops::BitOrAssign<&'a Bitboard> for Bitboard {
    fn bitor_assign(&mut self, rhs: &'a Bitboard) {
        self.p[0] |= rhs.p[0];
        self.p[1] |= rhs.p[1];
    }
}

impl<'a, 'b> ops::BitXor<&'a Bitboard> for &'b Bitboard {
    type Output = Bitboard;

    fn bitxor(self, rhs: &'a Bitboard) -> Bitboard {
        Bitboard { p: [self.p[0] ^ rhs.p[0], self.p[1] ^ rhs.p[1]] }
    }
}

impl<'a> ops::BitXorAssign<&'a Bitboard> for Bitboard {
    fn bitxor_assign(&mut self, rhs: &'a Bitboard) {
        self.p[0] ^= rhs.p[0];
        self.p[1] ^= rhs.p[1];
    }
}

impl<'a> ops::BitAnd<Square> for &'a Bitboard {
    type Output = Bitboard;

    fn bitand(self, rhs: Square) -> Bitboard {
        self & &SQUARE_BB[rhs.index()]
    }
}

impl<'a> ops::BitOr<Square> for &'a Bitboard {
    type Output = Bitboard;

    fn bitor(self, rhs: Square) -> Bitboard {
        self | &SQUARE_BB[rhs.index()]
    }
}

impl<'a> ops::BitXor<Square> for &'a Bitboard {
    type Output = Bitboard;

    fn bitxor(self, rhs: Square) -> Bitboard {
        self ^ &SQUARE_BB[rhs.index()]
    }
}

/////////////////////////////////////////////////////////////////////////////
// Trait implementation
/////////////////////////////////////////////////////////////////////////////

impl<'a> From<&'a Bitboard> for bool {
    fn from(bb: &'a Bitboard) -> bool {
        (bb.p[0] | bb.p[1]) != 0
    }
}

impl fmt::Display for Bitboard {
    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
        try!(writeln!(f, "   9   8   7   6   5   4   3   2   1"));
        try!(writeln!(f, "+---+---+---+---+---+---+---+---+---+"));

        for rank in 0..9 {
            try!(write!(f, "|"));
            for file in 0..9 {
                let sq = Square::new(file, rank).unwrap();
                try!(write!(f, " {} |", if bool::from(&(self & sq)) { "X" } else { " " }));
            }
            try!(writeln!(f, " {}", ('a' as u8 + rank) as char));
            try!(writeln!(f, "+---+---+---+---+---+---+---+---+---+"));
        }

        Ok(())
    }
}

/////////////////////////////////////////////////////////////////////////////
// Constants
/////////////////////////////////////////////////////////////////////////////

const SQUARE_BB: [Bitboard; 81] = [Bitboard { p: [1 << 0, 0] },
                                   Bitboard { p: [1 << 1, 0] },
                                   Bitboard { p: [1 << 2, 0] },
                                   Bitboard { p: [1 << 3, 0] },
                                   Bitboard { p: [1 << 4, 0] },
                                   Bitboard { p: [1 << 5, 0] },
                                   Bitboard { p: [1 << 6, 0] },
                                   Bitboard { p: [1 << 7, 0] },
                                   Bitboard { p: [1 << 8, 0] },
                                   Bitboard { p: [1 << 9, 0] },
                                   Bitboard { p: [1 << 10, 0] },
                                   Bitboard { p: [1 << 11, 0] },
                                   Bitboard { p: [1 << 12, 0] },
                                   Bitboard { p: [1 << 13, 0] },
                                   Bitboard { p: [1 << 14, 0] },
                                   Bitboard { p: [1 << 15, 0] },
                                   Bitboard { p: [1 << 16, 0] },
                                   Bitboard { p: [1 << 17, 0] },
                                   Bitboard { p: [1 << 18, 0] },
                                   Bitboard { p: [1 << 19, 0] },
                                   Bitboard { p: [1 << 20, 0] },
                                   Bitboard { p: [1 << 21, 0] },
                                   Bitboard { p: [1 << 22, 0] },
                                   Bitboard { p: [1 << 23, 0] },
                                   Bitboard { p: [1 << 24, 0] },
                                   Bitboard { p: [1 << 25, 0] },
                                   Bitboard { p: [1 << 26, 0] },
                                   Bitboard { p: [1 << 27, 0] },
                                   Bitboard { p: [1 << 28, 0] },
                                   Bitboard { p: [1 << 29, 0] },
                                   Bitboard { p: [1 << 30, 0] },
                                   Bitboard { p: [1 << 31, 0] },
                                   Bitboard { p: [1 << 32, 0] },
                                   Bitboard { p: [1 << 33, 0] },
                                   Bitboard { p: [1 << 34, 0] },
                                   Bitboard { p: [1 << 35, 0] },
                                   Bitboard { p: [1 << 36, 0] },
                                   Bitboard { p: [1 << 37, 0] },
                                   Bitboard { p: [1 << 38, 0] },
                                   Bitboard { p: [1 << 39, 0] },
                                   Bitboard { p: [1 << 40, 0] },
                                   Bitboard { p: [1 << 41, 0] },
                                   Bitboard { p: [1 << 42, 0] },
                                   Bitboard { p: [1 << 43, 0] },
                                   Bitboard { p: [1 << 44, 0] },
                                   Bitboard { p: [1 << 45, 0] },
                                   Bitboard { p: [1 << 46, 0] },
                                   Bitboard { p: [1 << 47, 0] },
                                   Bitboard { p: [1 << 48, 0] },
                                   Bitboard { p: [1 << 49, 0] },
                                   Bitboard { p: [1 << 50, 0] },
                                   Bitboard { p: [1 << 51, 0] },
                                   Bitboard { p: [1 << 52, 0] },
                                   Bitboard { p: [1 << 53, 0] },
                                   Bitboard { p: [1 << 54, 0] },
                                   Bitboard { p: [1 << 55, 0] },
                                   Bitboard { p: [1 << 56, 0] },
                                   Bitboard { p: [1 << 57, 0] },
                                   Bitboard { p: [1 << 58, 0] },
                                   Bitboard { p: [1 << 59, 0] },
                                   Bitboard { p: [1 << 60, 0] },
                                   Bitboard { p: [1 << 61, 0] },
                                   Bitboard { p: [1 << 62, 0] },
                                   Bitboard { p: [0, 1 << 0] },
                                   Bitboard { p: [0, 1 << 1] },
                                   Bitboard { p: [0, 1 << 2] },
                                   Bitboard { p: [0, 1 << 3] },
                                   Bitboard { p: [0, 1 << 4] },
                                   Bitboard { p: [0, 1 << 5] },
                                   Bitboard { p: [0, 1 << 6] },
                                   Bitboard { p: [0, 1 << 7] },
                                   Bitboard { p: [0, 1 << 8] },
                                   Bitboard { p: [0, 1 << 9] },
                                   Bitboard { p: [0, 1 << 10] },
                                   Bitboard { p: [0, 1 << 11] },
                                   Bitboard { p: [0, 1 << 12] },
                                   Bitboard { p: [0, 1 << 13] },
                                   Bitboard { p: [0, 1 << 14] },
                                   Bitboard { p: [0, 1 << 15] },
                                   Bitboard { p: [0, 1 << 16] },
                                   Bitboard { p: [0, 1 << 17] }];

const FILE_BB: [Bitboard; 9] = [Bitboard { p: [0x1FF << (9 * 0), 0] },
                                Bitboard { p: [0x1FF << (9 * 1), 0] },
                                Bitboard { p: [0x1FF << (9 * 2), 0] },
                                Bitboard { p: [0x1FF << (9 * 3), 0] },
                                Bitboard { p: [0x1FF << (9 * 4), 0] },
                                Bitboard { p: [0x1FF << (9 * 5), 0] },
                                Bitboard { p: [0x1FF << (9 * 6), 0] },
                                Bitboard { p: [0, 0x1FF << (9 * 0)] },
                                Bitboard { p: [0, 0x1FF << (9 * 1)] }];

const RANK_BB: [Bitboard; 9] = [Bitboard { p: [0x40201008040201 << 0, 0x201 << 0] },
                                Bitboard { p: [0x40201008040201 << 1, 0x201 << 1] },
                                Bitboard { p: [0x40201008040201 << 2, 0x201 << 2] },
                                Bitboard { p: [0x40201008040201 << 3, 0x201 << 3] },
                                Bitboard { p: [0x40201008040201 << 4, 0x201 << 4] },
                                Bitboard { p: [0x40201008040201 << 5, 0x201 << 5] },
                                Bitboard { p: [0x40201008040201 << 6, 0x201 << 6] },
                                Bitboard { p: [0x40201008040201 << 7, 0x201 << 7] },
                                Bitboard { p: [0x40201008040201 << 8, 0x201 << 8] }];

static ROOK_BLOCK_MASK: [Bitboard; 81] = [Bitboard { p: [0, 0] }; 81];

static BETWEEN_BB: [[Bitboard; 81]; 81] = [[Bitboard { p: [0, 0] }; 81]; 81];

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn square_bbs() {
        for i in 0..9 {
            println!("{}", FILE_BB[i]);
            println!("{}", RANK_BB[i]);
        }
    }
}